src/hubbard/hubbard_occupancies_derivatives.cpp

```diff
@@ -318,71 +315,93 @@ Hubbard::compute_occupancies_stress_derivatives(K_point&                    kp__
-                                                                               phi,
-                                                                               0, 0,
-                                                                               this->number_of_hubbard_orbitals());
-                        /* apply Q operator (diagonal in spin) */
-                        q_op__.apply(i, 0,
-                                     dphi, 0,
-                                     this->number_of_hubbard_orbitals(),
-                                     kp__.beta_projectors(),
-                                     dbeta_phi);
-                    }
-
-                    {
-                        /* non-collinear case */
-                        auto beta_phi = kp__.beta_projectors().inner<double_complex>(i,
-                                                                                     phi,
-                                                                                     0,
-                                                                                     0,
-                                                                                     this->number_of_hubbard_orbitals());
-                        /* apply Q operator (diagonal in spin) */
-                        q_op__.apply(i, 0, dphi, 0, this->number_of_hubbard_orbitals(), bp_strain_deriv, beta_phi);
-                    }
-                }
+            /* ths phitmp functions are non-magnetic at the moment, so copy spin-up channel to spin dn;
+             * this is done in order to have a gneral case of spin-dependent Hubbard orbitals */
+            if (ctx_.num_spins() == 2) {
+                phitmp.copy_from(device_t::CPU, phitmp.num_wf(), phitmp, 0, 0, 1, (ctx_.num_mag_dims() == 3) ? phitmp.num_wf() : 0);
             }
 
-            compute_occupancies(kp__,
-                                phi_s_psi,
-                                dphi_s_psi,
-                                dphi,
-                                dn__,
-                                dm,
-                                3 * nu + mu);
+            /* the rest of the code here has to be updated. Just to keep in mind:
+             *
+             * - kp.hubbard_wave_functions() are the (potentially) orthogonalized functions with S applied
+             * - kp.hubbard_atomic_wave_functions() are the non-orthofonal (initial) atomic Hubbard functions with S applied
+             * - both sets are spin-dependent, even if the orbitals have no spin label
+             * - atom_type.indexr_hub() is the new index of Hubbard radial functions
+             * - atom_type.indexb_hub() is the new index of Hubbard orbitals (radial function times a spherical harmonic)
+             * - wavefunctions_strain_deriv() and K_point::generate_atomic_wave_functions() work with the new indices
+             */
+
+            ////if (ctx_.processing_unit() == device_t::GPU) {
```

Comment:
@mtaillefumier Mathieu! I need a bit of your help from this line down to the end of the function. You can leave the GPU code for later, I can do it myself. 

|phitmp> contains d|phi>/d eps and S operator is not yet applied